### ISSUE TITLE
[QST] adata.X integer overflow for indptr

### ISSUE THREAD

**Issue by sdettle** [scverse/rapids_singlecell#249]
**Describe the bug**
After running UMAP on the GPUs adata seems fine, but is only about 2GB from the original ~35GB. The first couple thousand obs have values in the sparse matrix, but by then after that it is all zeros. I have tried multiple methods of loading and unloading the data onto the GPU and Back onto the cpu. The input adata ('In_Vivo_NHP_and_Human_10min.h5ad') is perfect. I am running a pretty simplistic pipeline at this point, so I won't be putting in example data just my workflow. 

The original adata.X is {<728593x16413 sparse matrix of type '<class 'numpy.float64'>'
	with 2199112212 stored elements in Compressed Sparse Row format>}

The returned adata.X after umap is {<728593x16413 sparse matrix of type '<class 'numpy.float64'>'
	with -2095855084 stored elements in Compressed Sparse Row format>}

The UMAP and the Leiden Clustering works and in lickety split fashion(not shown)
![umapumap_cd34](https://github.com/user-attachments/assets/cb683547-b6ae-40f9-bfcb-a47043d47dd6)

Finally, I don't expect this to be a ram issue as I have around 200gB and the uncompressed h5ad is around ~35Gb

For now I will just extract the obsm and the UNS and then append to the original h5ad, but this seems to be concerning behavior with no errors raised it took me a long time to figure what was causing downstream code to crash.

**Expected behavior**
adata.X shouldn't change after saving as nothing being changed on the .X

**Environment details (please complete the following information):**
 - Environment location: [bare-metal on the cloud. The only way I could install the rsc was using PyPI in the base environment]
 - Linux Distro/Architecture: [Debian 12 amd64]
 - GPU Model/Driver: [Tesla T4 560.28.03              Driver Version: 560.28.03    ]
 - CUDA: [ CUDA Version: 12.6]
 - Method of Rapids install: [pip and nvidia/cuda from source ]
absl-py                   2.1.0
adal                      1.2.7
adata                     2.5.7
aiohappyeyeballs          2.3.7
aiohttp                   3.10.4
aiosignal                 1.3.1
anndata                   0.10.8
anyio                     4.4.0
archspec                  0.2.3
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
array_api_compat          1.8
arrow                     1.3.0
asttokens                 2.0.5
attrs                     24.2.0
backcall                  0.2.0
beautifulsoup4            4.12.3
bleach                    6.1.0
bokeh                     3.4.3
boltons                   23.0.0
Brotli                    1.0.9
cachetools                5.4.0
certifi                   2024.7.4
cffi                      1.16.0
charset-normalizer        2.0.4
chex                      0.1.86
click                     8.1.7
cloudauthz                0.6.0
cloudpickle               3.0.0
colorcet                  3.1.0
comm                      0.1.2
conda                     24.7.1
conda-content-trust       0.2.0
conda-libmamba-solver     23.12.0
conda-package-handling    2.2.0
conda_package_streaming   0.9.0
contextlib2               21.6.0
contourpy                 1.2.1
cryptography              41.0.7
cucim-cu12                24.8.0
cuda-python               12.6.0
cudf-cu12                 24.8.2
cugraph-cu12              24.8.0
cuml-cu12                 24.8.0
cuproj-cu12               24.8.0
cupy-cuda12x              13.2.0
cuspatial-cu12            24.8.0
cuvs-cu12                 24.8.0
cuxfilter-cu12            24.8.0
cycler                    0.12.1
dask                      2024.7.1
dask-cuda                 24.8.2
dask-cudf-cu12            24.8.2
dask-expr                 1.1.9
datashader                0.16.3
debugpy                   1.6.7
decorator                 5.1.1
defusedxml                0.7.1
distinctipy               1.3.4
distributed               2024.7.1
distributed-ucxx-cu12     0.39.1
distro                    1.8.0
docrep                    0.3.2
etils                     1.9.2
executing                 0.8.3
fastjsonschema            2.20.0
fastrlock                 0.8.2
filelock                  3.15.4
flax                      0.8.5
fonttools                 4.53.1
fqdn                      1.5.1
frozendict                2.4.2
frozenlist                1.4.1
fsspec                    2024.6.1
geopandas                 1.0.1
h5py                      3.11.0
holoviews                 1.19.1
idna                      3.4
imageio                   2.35.0
importlib_metadata        8.2.0
importlib_resources       6.4.3
ipykernel                 6.25.0
ipython                   8.15.0
isoduration               20.11.0
jax                       0.4.31
jax-cuda12-pjrt           0.4.31
jax-cuda12-plugin         0.4.31
jaxlib                    0.4.31
jedi                      0.18.1
Jinja2                    3.1.4
joblib                    1.4.2
jsonpatch                 1.32
jsonpointer               2.1
jsonschema                4.23.0
jsonschema-specifications 2023.12.1
jupyter_client            8.6.0
jupyter_core              5.5.0
jupyter-events            0.10.0
jupyter_server            2.14.2
jupyter_server_proxy      4.3.0
jupyter_server_terminals  0.5.3
jupyterlab_pygments       0.3.0
kiwisolver                1.4.5
lazy_loader               0.4
legacy-api-wrap           1.4
libmambapy                1.5.3
libucx-cu12               1.15.0.post1
lightning                 2.1.4
lightning-utilities       0.11.6
linkify-it-py             2.0.3
llvmlite                  0.43.0
locket                    1.0.0
Markdown                  3.7
markdown-it-py            3.0.0
MarkupSafe                2.1.5
matplotlib                3.9.2
matplotlib-inline         0.1.6
mdit-py-plugins           0.4.1
mdurl                     0.1.2
menuinst                  2.0.1
mistune                   3.0.2
ml-collections            0.1.1
ml-dtypes                 0.4.0
mpmath                    1.3.0
msgpack                   1.0.8
mudata                    0.3.0
multidict                 6.0.5
multipledispatch          1.0.0
natsort                   8.4.0
nbclient                  0.10.0
nbconvert                 7.16.4
nbformat                  5.10.4
nest-asyncio              1.5.6
networkx                  3.3
numba                     0.60.0
numpy                     1.26.4
numpyro                   0.15.2
nvidia-cublas-cu12        12.1.3.1
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvcc-cu12     12.6.20
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu12  12.1.105
nvidia-cudnn-cu12         9.1.0.70
nvidia-cufft-cu12         11.0.2.54
nvidia-curand-cu12        10.3.2.106
nvidia-cusolver-cu12      11.4.5.107
nvidia-cusparse-cu12      12.1.0.106
nvidia-nccl-cu12          2.20.5
nvidia-nvjitlink-cu12     12.6.20
nvidia-nvtx-cu12          12.1.105
nvtx                      0.2.10
OpenMM                    8.1.2
opt-einsum                3.3.0
optax                     0.2.3
orbax-checkpoint          0.5.23
overrides                 7.7.0
packaging                 23.1
pandas                    2.2.2
pandocfilters             1.5.1
panel                     1.4.5
param                     2.1.1
parso                     0.8.3
partd                     1.4.2
patsy                     0.5.6
pdbfixer                  1.9.0
pexpect                   4.8.0
pickleshare               0.7.5
pillow                    10.4.0
pip                       23.3.1
platformdirs              3.10.0
pluggy                    1.0.0
prometheus_client         0.20.0
prompt-toolkit            3.0.36
protobuf                  5.27.3
psutil                    5.9.0
ptyprocess                0.7.0
pure-eval                 0.2.2
py-mini-racer             0.6.0
pyarrow                   16.1.0
pycosat                   0.6.6
pycparser                 2.21
pyct                      0.5.0
Pygments                  2.15.1
PyJWT                     2.9.0
pylibcugraph-cu12         24.8.0
pylibraft-cu12            24.8.1
pynndescent               0.5.13
pynvjitlink-cu12          0.3.0
pynvml                    11.4.1
pyogrio                   0.9.0
pyOpenSSL                 23.2.0
pyparsing                 3.1.2
pyproj                    3.6.1
pyro-api                  0.1.2
pyro-ppl                  1.9.1
PySocks                   1.7.1
python-dateutil           2.8.2
python-json-logger        2.0.7
pytorch-lightning         2.4.0
pytz                      2024.1
pyviz_comms               3.0.3
PyYAML                    6.0.2
pyzmq                     25.1.0
raft-dask-cu12            24.8.1
rapids-dask-dependency    24.8.0
rapids_singlecell         0.10.8
referencing               0.35.1
requests                  2.31.0
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.7.1
rmm-cu12                  24.8.2
rpds-py                   0.20.0
ruamel.yaml               0.17.21
scanpy                    1.10.2
scikit-image              0.23.2
scikit-learn              1.5.1
scikit-misc               0.5.1
scipy                     1.13.1
scvi-tools                1.1.5
seaborn                   0.13.2
Send2Trash                1.8.3
session_info              1.0.0
setuptools                68.2.2
shapely                   2.0.5
simpervisor               1.0.0
six                       1.16.0
sniffio                   1.3.1
sortedcontainers          2.4.0
soupsieve                 2.6
stack-data                0.2.0
statsmodels               0.14.2
stdlib-list               0.10.0
sympy                     1.13.2
tblib                     3.0.0
tensorstore               0.1.64
terminado                 0.18.1
threadpoolctl             3.5.0
tifffile                  2024.8.10
tinycss2                  1.3.0
toolz                     0.12.1
torch                     2.4.0
torchmetrics              1.4.1
tornado                   6.3.3
tqdm                      4.65.0
traitlets                 5.7.1
treelite                  4.3.0
triton                    3.0.0
truststore                0.8.0
types-python-dateutil     2.9.0.20240316
typing_extensions         4.12.2
tzdata                    2024.1
uc-micro-py               1.0.3
ucx-py-cu12               0.39.2
ucxx-cu12                 0.39.1
umap-learn                0.5.6
uri-template              1.3.0
urllib3                   1.26.18
wcwidth                   0.2.5
webcolors                 24.8.0
webencodings              0.5.1
websocket-client          1.8.0
wheel                     0.41.2
xarray                    2024.7.0
xyzservices               2024.6.0
yarl                      1.9.4
zict                      3.0.0
zipp                      3.20.0
zstandard                 0.19.0
   - If method of install is [from source], provide versions of `cmake` & `gcc/g++` and commit hash of build

**Additional context**

'''
python

#run in base
#rsc-scvi-umap-nhp-hu-vivo.py

import rapids_singlecell as rsc
import anndata as ad
import scvi
#import cupyx
import rmm
import cupy as cp
from rmm.allocators.cupy import rmm_cupy_allocator

rmm.reinitialize(
    managed_memory=True,
    pool_allocator=False,
)
cp.cuda.set_allocator(rmm_cupy_allocator)

output_dir = '/home/sdettle/remote_jupyter_work/Human_NHP_Fetal_Hemato/h5ads_post_qc/'
file_name = output_dir + 'In_Vivo_NHP_and_Human_10min.h5ad'

adata = ad.read_h5ad(
            file_name, 
            chunk_size=10000000, 
            #backed = None
)

model_dir = output_dir + 'In_Vivo_NHP_and_Human_10min_scvi_model'

scvi.model.SCVI.setup_anndata(
    adata,
    batch_key = 'BioProject',
    categorical_covariate_keys=['Technology', 'Species'],
    continuous_covariate_keys=["n_counts", "n_genes", 'total_counts']
)

model = scvi.model.SCVI.load(model_dir, adata=adata)

SCVI_LATENT_KEY = "Latent_SCVI_Vivo_Hu_NH"
latent = model.get_latent_representation()
adata.obsm[SCVI_LATENT_KEY] = latent

rsc.get.anndata_to_GPU(adata)
rsc.pp.neighbors(adata, use_rep=SCVI_LATENT_KEY, key_added = 'vivo_nhp_human_neighbors_scvi')

print('umap')
rsc.tl.umap(adata, min_dist=0.3, neighbors_key = 'vivo_nhp_human_neighbors_scvi')

file_name_save = output_dir + 'In_Vivo_NHP_and_Human_10min_scvi_umap_pds.h5ad'
rsc.get.anndata_to_CPU(adata)
print(f'saving to {file_name_save}')
adata.write_h5ad(
    file_name_save,
    #as_dense = 'X'
)

)

'''

**Comment by sdettle** [comment]
Hey, I appreciate that. I fixed it by commenting out the rsc.get.anndata_to_GPU(adata)  and also the rsc.get.anndata_to_CPU(adata). I'll figure out how I'll convert my sparse matrix to int32 if I need to in the future, but most of my functions I use in scvi/scanpy (works in jupyter for me (conda)) , but the rapids cells' umap function is insanely fast. Saved me from days of compute to mere seconds (though installing cuda and nvidia drivers filled the amount of time I woulda saved (3+ days)) . Thanks for all yalls work on this.
